flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java

```diff
@@ -0,0 +1,1667 @@
+        assertor.shouldEmit(
+                testHarness,
+                rowOfKind(
+                        RowKind.INSERT,
+                        "Ord#1",
+                        "LineOrd#2",
+                        "4 Bellevue Drive, Pottstown, PB 19464",
+                        "Ord#1",
+                        "LineOrd#2",
+                        "SHIP"),
+                rowOfKind(
+                        RowKind.INSERT,
+                        "Ord#1",
+                        "LineOrd#2",
+                        "4 Bellevue Drive, Pottstown, PB 19464",
+                        "Ord#1",
+                        "LineOrd#2",
+                        "SHIP"));
+        testHarness.processElement1(
+                updateBeforeRecord("Ord#1", "LineOrd#2", "4 Bellevue Drive, Pottstown, PB 19464"));
+        testHarness.processElement1(
+                updateAfterRecord("Ord#1", "LineOrd#2", "4 Bellevue Drive, Pottstown, PB 19464"));
+        testHarness.processElement2(insertRecord("Ord#1", "LineOrd#3", "AIR"));
+        testHarness.processElement1(
+                deleteRecord("Ord#6", "LineOrd#1", "8 Bellevue Drive, Pottstown, PF 19464"));
+        testHarness.processElement2(deleteRecord("Ord#1", "LineOrd#2", "SHIP"));
+        // right state
+        // "Ord#1", "LineOrd#1", "AIR"
+        // "Ord#1", "LineOrd#3", "AIR"
+        testHarness.processElement1(
+                insertRecord("Ord#1", "LineOrd#2", "4 Bellevue Drive, Pottstown, PB 19464"));
+        assertor.shouldEmit(
+                testHarness,
+                rowOfKind(
+                        RowKind.INSERT,
+                        "Ord#3",
+                        "LineOrd#3",
+                        "5 Bellevue Drive, Pottstown, PD 19464",
+                        "Ord#1",
+                        "LineOrd#3",
+                        "AIR"),
+                rowOfKind(
+                        RowKind.DELETE,
+                        "Ord#1",
+                        "LineOrd#2",
+                        "4 Bellevue Drive, Pottstown, PB 19464",
+                        "Ord#1",
+                        "LineOrd#2",
+                        "SHIP"),
+                rowOfKind(
+                        RowKind.DELETE,
+                        "Ord#1",
+                        "LineOrd#2",
+                        "4 Bellevue Drive, Pottstown, PB 19464",
+                        "Ord#1",
+                        "LineOrd#2",
+                        "SHIP"),
+                rowOfKind(
+                        RowKind.DELETE,
+                        "Ord#6",
+                        "LineOrd#1",
+                        "8 Bellevue Drive, Pottstown, PF 19464",
+                        "Ord#1",
+                        "LineOrd#1",
+                        "AIR"));
+    }
+
+    /** Outer join only emits INSERT or DELETE Msg. */
+    @Tag("miniBatchSize=10")
+    @Test
+    public void testLeftJoinWithJoinKeyContainsUniqueKey() throws Exception {
```

Comment:
I think there should be some more tests for LEFT JOIN, including logic for record folding and suppression, mainly for some boundary cases.

Reply:
Ok, I think update records with more rowkinds and disorder could be tested. I would improve the coverage of tests.